compiler/sem/sigmatch.nim

```diff
@@ -1597,88 +1593,70 @@ typeRel can be used to establish various relationships between types:
         result = typeRel(c, ff, a, flags)
 
   of tyGenericInvocation:
-    var x = a.skipGenericAlias
-    let concpt = f[0].skipTypes({tyGenericBody})
-
-    # XXX: This is very hacky. It should be moved back into liftTypeParam
-    if x.kind in {tyGenericInst, tyArray} and
-      c.calleeSym != nil and
-      c.calleeSym.kind in {skProc, skFunc} and c.call != nil:
-      let inst = prepareMetatypeForSigmatch(c.c, c.bindings, c.call.info, f)
-      #echo "inferred ", typeToString(inst), " for ", f
-      return typeRel(c, inst, a, flags)
+    # a generic invocation represents an unresolved type application
+    let
+      x = a.skipGenericAlias
+      # XXX: ^^ this means argument phantom types are ignored
+      body = f.base.lastSon
+    var sp: SkippedPtr
 
     if x.kind == tyGenericInvocation:
-      if f[0] == x[0]:
+      # an unresolved type application is matched against another unresolved
+      # type application
+      if f.base == x.base:
         for i in 1..<f.len:
           let tr = typeRel(c, f[i], x[i], flags)
           if tr <= isSubtype: return
         result = isGeneric
-    elif x.kind == tyGenericInst and f[0] == x[0] and
-          x.len - 1 == f.len:
-      for i in 1..<f.len:
-        c.c.graph.config.internalAssert(x[i].kind != tyGenericParam, "wrong instantiated type!")
-
-        if typeRel(c, f[i], x[i], flags) <= isSubtype:
-          # Workaround for https://github.com/nim-lang/nim/issues/4589
-          if f[i].kind != tyTypeDesc:
-            return
-
+      # XXX: what about nested invocations (i.e., generic aliases)?
+    elif x.kind == tyGenericInst and f.base == x.base:
+      # a resolved type application is matched against an unresolved one where
+      # both are for the same generic type -> the arguments have to match
+      result = compareInvocationArguments(c, f, x, flags)
+    elif x.kind == tyGenericBody and f.base == x:
+      # this is a special case to support:
+      #   proc f[T](a: typedesc[Generic[T]])
+      #   f[int](Generic)
+      # XXX: this seems like a really bad idea, why should that work? The
+      #      procedure wants a typedesc for an instantiated type, but it gets
+      #      an (unlifted) composite-type-class
```

Comment:
It shouldn't, pretty sure it should be an error where `Generic[int]` is expected, but received a `Generic`.

Specifying the type parameter on the procedure should affect the formal parameter's type definition and interpretation of any argument (the instantiated procedure's declaration), but not the actual argument.